### ISSUE TITLE
pathlib copy methods do not work across partitions

### ISSUE THREAD

**Issue by Akuli** [python/cpython#122890]
# Bug report

### Bug description:

On linux, after compiling latest `main` branch:

```
$ ./python 
Python 3.14.0a0 (heads/main:5580f31c56e, Aug 10 2024, 22:50:58) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pathlib
>>> pathlib.Path("python").copy("/tmp/foo")
Traceback (most recent call last):
  ...
OSError: [Errno 18] Invalid cross-device link: 'python' -> '/tmp/foo'
>>> pathlib.Path("Lib").copytree("/tmp/bar")
Traceback (most recent call last):
  ...
OSError: [Errno 18] Invalid cross-device link: 'Lib/ast.py' -> '/tmp/bar/ast.py'
>>> 
```

On this computer, `/home` and `/tmp` are on different partitions. If I copy within the same partition (e.g. `./python` to `/home/akuli/foo`), that works.

### CPython versions tested on:

CPython main branch

### Operating systems tested on:

Linux

### Linked PRs
* gh-122897

**Comment by eryksun** [comment]
`pathlib.Path.copy()` tries to use `copy_file_range()`, but it doesn't handle an `EXDEV` error. Support for cross-filesystem copies was introduced in Linux 5.3, but since Linux 5.19 (and 5.15 LTS), this capability has been limited to cases in which both filesystems are of the same type, and the filesystem explicitly supports `copy_file_range()`.

**Comment by barneygale** [comment]
Thanks @Akuli for testing the new methods and reporting the issue, and @eryksun for the analysis. I've opened a PR that should fix this: #122897

**Comment by barneygale** [comment]
Fixed!
